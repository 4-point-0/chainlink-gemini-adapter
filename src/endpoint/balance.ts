import { AdapterEndpoint } from "@chainlink/external-adapter-framework/adapter";
import { InputParameters } from "@chainlink/external-adapter-framework/validation";
import { config } from "../config";
import { balanceTransport } from "../transport/balance";
import { AdapterRequest } from "@chainlink/external-adapter-framework/util";
import { AdapterInputError } from "@chainlink/external-adapter-framework/validation/error";
import { ethers } from "ethers";

export const inputParameters = new InputParameters(
  {
    addresses: {
      required: true,
      description: "array of object",
      array: true,
      type: {
        address: {
          required: true,
          description: "eth address",
          type: "string",
        },
        blockHeight: {
          required: false,
          description: "block height. default - latest",
          type: "number",
        },
      },
    },
  },
  [
    {
      addresses: [
        { address: "0x103B66487784F6e3b4C5B2AcA92758198554C3e1" },
        {
          address: "0x0AA2ca7B13855e115EBdCDA89E4f19EC2c1181E5",
          blockHeight: 100002000,
        },
      ],
    },
  ]
);

export type BaseEndpointTypes = {
  Parameters: typeof inputParameters.definition;
  Response: {
    Result: string;
    Data: {
      result: { address: string; balance: string }[];
    };
  };
  Settings: typeof config.settings;
};

export const endpoint = new AdapterEndpoint({
  name: "balance",
  aliases: [],
  transport: balanceTransport,
  inputParameters,
  customInputValidation: (
    req: AdapterRequest<typeof inputParameters.validated>
  ): AdapterInputError | undefined => {
    const addresses = req.requestContext.data.addresses;
    if (req.requestContext.data.addresses.length === 0) {
      throw new AdapterInputError({
        statusCode: 400,
        message: `Input, at 'addresses', must be a non-empty array of objects.`,
      });
    }
    for (const address of addresses) {
      if (!ethers.utils.isAddress(address.address)) {
        throw new AdapterInputError({
          statusCode: 400,
          message: `Invalid Ethereum address: ${address.address}`,
        });
      }
      if (
        address.blockHeight !== undefined &&
        (address.blockHeight < 0 ||
          address.blockHeight > Number.MAX_SAFE_INTEGER)
      ) {
        throw new AdapterInputError({
          statusCode: 400,
          message: `Invalid blockHeight for address ${address.address}: ${address.blockHeight}. It must be a non-negative integer.`,
        });
      }
    }
    return;
  },
});
