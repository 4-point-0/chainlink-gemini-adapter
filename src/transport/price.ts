import { WebsocketReverseMappingTransport } from "@chainlink/external-adapter-framework/transports";
import { BaseEndpointTypes, inputParameters } from "../endpoint/price";
import { makeLogger } from "@chainlink/external-adapter-framework/util";

const logger = makeLogger("Price Websocket Endpoint");

export interface WSResponse {
  changes: [[string, string, string]];
  type: string;
  symbol: string;
  result: string;
  reason: string;
}

export type WsTransportTypes = BaseEndpointTypes & {
  Provider: {
    WsMessage: WSResponse;
  };
};

let latestParams: typeof inputParameters.validated = { base: "", quote: "" };
export const wsTransport: WebsocketReverseMappingTransport<
  WsTransportTypes,
  string
> = new WebsocketReverseMappingTransport<WsTransportTypes, string>({
  url: (context) => context.adapterSettings.WS_API_ENDPOINT,
  handlers: {
    message(message) {
      if (message.result === "error") {
        logger.error(`Data provider Error: ${JSON.stringify(message)}`);
        return [
          {
            params: latestParams,
            response: {
              statusCode: 502,
              errorMessage: message.reason,
            },
          },
        ];
      }

      const changesType = message.changes[0][0];
      if (changesType !== "sell") {
        return;
      }

      const price = parseFloat(message.changes[0][1]);
      const params: ({ base: string; quote: string } & {}) | undefined =
        wsTransport.getReverseMapping(message.symbol);

      if (!params) {
        logger.error(
          "The Data Provider didn't return expected Gemini pair symbol"
        );
        return;
      }

      return [
        {
          params: params,
          response: {
            statusCode: 200,
            result: price,
            data: {
              result: price,
            },
            timestamps: {
              providerIndicatedTimeUnixMs: Date.now(), // Gemini doesn't provide a timestamp in the message
            },
          },
        },
      ];
    },
  },
  builders: {
    subscribeMessage: (params) => {
      latestParams = params;
      const pair = `${params.base}${params.quote}`.toUpperCase();
      wsTransport.setReverseMapping(pair, params);
      return JSON.stringify({
        type: "subscribe",
        subscriptions: [
          {
            name: "l2",
            symbols: [pair],
          },
        ],
      });
    },
    unsubscribeMessage: (params) => {
      return JSON.stringify({
        type: "unsubscribe",
        subscriptions: [
          {
            name: "l2",
            symbols: [`${params.base}${params.quote}`.toUpperCase()],
          },
        ],
      });
    },
  },
});
