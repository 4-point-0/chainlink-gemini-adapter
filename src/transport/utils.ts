import { AdapterResponse } from "@chainlink/external-adapter-framework/util";
import { BigNumber, BigNumberish, ethers } from "ethers";
import { BalanceTransportTypes } from "./balance";

export const formatAmount = (amount: BigNumberish, toFixed = 2) => {
  return parseFloat(ethers.utils.formatUnits(amount, "ether"))
    .toFixed(toFixed)
    .toString();
};

export const makeResponse = (
  balances: BigNumber[],
  results: {
    address: any;
    balance: string;
  }[]
) => {
  const response: AdapterResponse<BalanceTransportTypes["Response"]> = {
    result: formatAmount(
      balances.reduce((sum, balance) => sum.add(balance), BigNumber.from(0))
    ),
    statusCode: 200,
    data: { result: results },
    timestamps: {
      providerDataRequestedUnixMs: Date.now(),
      providerDataReceivedUnixMs: Date.now(),
      providerIndicatedTimeUnixMs: undefined,
    },
  };
  return response;
};

export const getBalances = async (
  addresses: ({
    address: string;
  } & {
    blockHeight?: number | undefined;
  })[],
  provider: ethers.providers.JsonRpcProvider
) => {
  const balances = await Promise.all(
    addresses.map((addr: any) =>
      provider.getBalance(
        addr.address.toLowerCase(),
        addr.blockHeight ?? "latest"
      )
    )
  );

  const results = addresses.map((addr: any, index: number) => ({
    address: addr.address,
    balance: formatAmount(balances[index]._hex),
  }));

  return { balances, results };
};
