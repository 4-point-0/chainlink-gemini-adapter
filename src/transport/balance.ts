import {
  Transport,
  TransportDependencies,
} from "@chainlink/external-adapter-framework/transports";
import { BaseEndpointTypes, inputParameters } from "../endpoint/balance";
import { ethers } from "ethers";
import { getBalances, makeResponse } from "./utils";
import {
  AdapterRequest,
  AdapterResponse,
} from "@chainlink/external-adapter-framework/util";
import { ResponseCache } from "@chainlink/external-adapter-framework/cache/response";
import { AdapterError } from "@chainlink/external-adapter-framework/validation/error";
import { config } from "../config";
import { EndpointContext } from "@chainlink/external-adapter-framework/adapter";

interface Result {
  address: string;
  balance: string;
}

export interface ResponseSchema {
  data: {
    result: Result[];
  };
}

export type BalanceTransportTypes = BaseEndpointTypes & {
  Provider: {
    RequestBody: never;
    ResponseBody: ResponseSchema;
  };
};

interface AddressInfo {
  address: string;
  blockHeight?: number;
}

export class LimitedCapacitySet<T> {
  private items: T[] = [];
  private itemSet: Set<T> = new Set();
  private capacity: number;

  constructor(capacity: number) {
    this.capacity = capacity;
  }

  add(item: T): void {
    if (!this.itemSet.has(item)) {
      if (this.items.length >= this.capacity) {
        const oldestItem = this.items.shift();
        if (oldestItem) this.itemSet.delete(oldestItem);
      }
      this.items.push(item);
      this.itemSet.add(item);
    }
  }

  has(item: T): boolean {
    return this.itemSet.has(item);
  }

  delete(item: T): void {
    const index = this.items.indexOf(item);
    if (index > -1) {
      this.items.splice(index, 1);
      this.itemSet.delete(item);
    }
  }

  values(): IterableIterator<T> {
    return this.itemSet.values();
  }
}

export class BalanceTransport implements Transport<BalanceTransportTypes> {
  name!: string;
  responseCache!: ResponseCache<BalanceTransportTypes>;
  provider!: ethers.providers.JsonRpcProvider;
  currentRequests = new LimitedCapacitySet<string>(10000); // capacity of this.responseCache.cache is 10000
  listener: ethers.providers.Listener | null = null;

  async initialize(
    dependencies: TransportDependencies<BalanceTransportTypes>,
    settings: typeof config.settings,
    _endpointName: string,
    transportName: string
  ): Promise<void> {
    this.responseCache = dependencies.responseCache;
    this.name = transportName;
    this.provider = new ethers.providers.JsonRpcProvider(
      settings.ETHEREUM_RPC_URL,
      settings.CHAIN_ID
    );
  }

  async registerRequest(
    req: AdapterRequest<typeof inputParameters.validated>,
    _adapterSettings: BalanceTransportTypes["Settings"]
  ): Promise<void> {
    const addresses: typeof inputParameters.validated.addresses =
      req.requestContext.data.addresses;
    const requestId = this.getRequestId(addresses);
    this.currentRequests.add(requestId);
  }

  private getRequestId(addresses: AddressInfo[]): string {
    return JSON.stringify(addresses);
  }

  async backgroundExecute(
    _context: EndpointContext<BalanceTransportTypes>
  ): Promise<void> {
    if (this.listener) return;

    this.listener = async () => {
      for (const requestId of this.currentRequests.values()) {
        const addresses: typeof inputParameters.validated.addresses =
          JSON.parse(requestId);
        try {
          const { balances, results } = await getBalances(
            addresses,
            this.provider
          );

          const response = makeResponse(balances, results);

          await this.responseCache.write(this.name, [
            {
              params: {
                addresses: addresses.map((address: any) => ({
                  address: address.address,
                  blockHeight: address.blockHeight,
                })),
              },
              response,
            },
          ]);
        } catch (e: any) {
          console.error(
            `Error updating balances for request ${requestId}:`,
            e.message
          );
        }
      }
    };

    this.provider.on("block", this.listener);
  }

  async foregroundExecute(
    req: AdapterRequest<typeof inputParameters.validated>
  ): Promise<AdapterResponse<BalanceTransportTypes["Response"]>> {
    const cachedResponse = await this.responseCache.cache.get(
      req.requestContext.cacheKey
    );

    if (cachedResponse) {
      return cachedResponse;
    }

    const addresses: typeof inputParameters.validated.addresses =
      req.requestContext.data.addresses;

    try {
      const { balances, results } = await getBalances(addresses, this.provider);
      const response = makeResponse(balances, results);
      await this.responseCache.write(this.name, [
        {
          params: {
            addresses: addresses.map((address: any) => ({
              address: address.address,
              blockHeight: address.blockHeight,
            })),
          },
          response,
        },
      ]);

      return response;
    } catch (error) {
      throw new AdapterError({
        statusCode: 502,
        message: `Error fetching balances: ${(error as Error).message}`,
      });
    }
  }
}

export const balanceTransport = new BalanceTransport();
