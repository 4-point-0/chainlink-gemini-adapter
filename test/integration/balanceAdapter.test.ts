import {
  TestAdapter,
  setEnvVariables,
} from "@chainlink/external-adapter-framework/util/testing-utils";
import * as nock from "nock";
import { mockRpcEndpoint, mockBalances, mockChainId } from "./fixtures";
import { LimitedCapacitySet } from "../../src/transport/balance";

jest.mock("ethers", () => {
  const actualModule = jest.requireActual("ethers");
  return {
    ...actualModule,
    ethers: {
      ...actualModule.ethers,
      providers: {
        JsonRpcProvider: function () {
          return {
            getBalance: jest.fn().mockImplementation((address) => {
              return mockBalances[address];
            }),
            on: jest.fn().mockImplementation((event, listener) => {
              if (event === "block") {
                setTimeout(() => listener(), 0);
              }
              return {
                removeListener: jest.fn(),
              };
            }),
            removeListener: jest.fn(),
          };
        },
      },
    },
  };
});

describe("execute", () => {
  let spy: jest.SpyInstance;
  let testAdapter: TestAdapter;
  let oldEnv: NodeJS.ProcessEnv;
  const rpcUrl = mockRpcEndpoint;
  const chainId = mockChainId;

  beforeAll(async () => {
    oldEnv = JSON.parse(JSON.stringify(process.env));

    process.env.ETHEREUM_RPC_URL = rpcUrl;
    process.env.CHAIN_ID = chainId;

    const mockDate = new Date("2001-01-01T11:11:11.111Z");
    spy = jest.spyOn(Date, "now").mockReturnValue(mockDate.getTime());

    const adapter = (await import("../../src")).adapter;
    adapter.rateLimiting = undefined;
    testAdapter = await TestAdapter.startWithMockedCache(adapter, {
      testAdapter: {} as TestAdapter<never>,
    });
  });

  afterAll(async () => {
    setEnvVariables(oldEnv);
    await testAdapter.api.close();
    nock.restore();
    nock.cleanAll();
    spy.mockRestore();
  });

  describe("balance endpoint", () => {
    it("should return success for a single address", async () => {
      const data = {
        endpoint: "balance",
        addresses: [{ address: "0x103b66487784f6e3b4c5b2aca92758198554c3e1" }],
      };
      const response = await testAdapter.request(data);
      const body = JSON.parse(response.body);

      expect(body.statusCode).toBe(200);
      expect(body.data.result[0].balance).toBe("14.61");
      expect(body.result).toBe("14.61");
      expect(response.json()).toMatchSnapshot();
    }, 20000);

    it("should return success for multiple addresses", async () => {
      const data = {
        endpoint: "balance",
        addresses: [
          {
            address: "0x103b66487784f6e3b4c5b2aca92758198554c3e1",
          },
          {
            address: "0x742d35cc6634c0532925a3b844bc454e4438f44e",
            blockHeight: 20540144,
          },
        ],
      };
      const response = await testAdapter.request(data);

      const body = JSON.parse(response.body);
      expect(body.statusCode).toBe(200);
      expect(body.data.result[0].balance).toBe("14.61");
      expect(body.data.result[1].balance).toBe("0.78");
      expect(body.result).toBe("15.40");
      expect(response.json()).toMatchSnapshot();
    });

    it("should handle invalid address error responses", async () => {
      const data = {
        endpoint: "balance",
        addresses: [{ address: "0xInvalidAddress" }],
      };
      const response = await testAdapter.request(data);
      const body = JSON.parse(response.body);

      expect(body.statusCode).toBe(400);
      expect(body.error.message).toBe(
        "Invalid Ethereum address: 0xInvalidAddress"
      );
      expect(response.json()).toMatchSnapshot();
    });

    it("should handle invalid blockheight error responses", async () => {
      const data = {
        endpoint: "balance",
        addresses: [
          {
            address: "0x103B66487784F6e3b4C5B2AcA92758198554C3e1",
            blockHeight: -1,
          },
        ],
      };
      const response = await testAdapter.request(data);
      const body = JSON.parse(response.body);

      expect(body.statusCode).toBe(400);
      expect(body.error.message).toBe(
        "Invalid blockHeight for address 0x103B66487784F6e3b4C5B2AcA92758198554C3e1: -1. It must be a non-negative integer."
      );
      expect(response.json()).toMatchSnapshot();
    });

    it("should handle empty address list", async () => {
      const data = {
        endpoint: "balance",
        addresses: [],
      };
      const response = await testAdapter.request(data);
      const body = JSON.parse(response.body);
      expect(body.statusCode).toBe(400);
      expect(body.error.message).toBe(
        "Input, at 'addresses', must be a non-empty array of objects."
      );
      expect(response.json()).toMatchSnapshot();
    });

    it("should maintain a maximum of 10000 items", () => {
      const currentRequests = new LimitedCapacitySet<string>(10000);

      const addresses = Array.from({ length: 10100 }, (_, i) => ({
        address: `0x${i.toString().padStart(40, "0")}`,
      }));

      for (const addr of addresses) {
        const requestId = JSON.stringify([addr]);
        currentRequests.add(requestId);
      }

      expect(Array.from(currentRequests.values()).length).toBe(10000);

      for (let i = 0; i < 100; i++) {
        const oldestAddr = `0x${i.toString().padStart(40, "0")}`;
        const requestId = JSON.stringify([{ address: oldestAddr }]);
        expect(currentRequests.has(requestId)).toBe(false);
      }

      for (let i = 100; i < 10100; i++) {
        const newestAddr = `0x${i.toString().padStart(40, "0")}`;
        const requestId = JSON.stringify([{ address: newestAddr }]);
        expect(currentRequests.has(requestId)).toBe(true);
      }
    });
  });
});
