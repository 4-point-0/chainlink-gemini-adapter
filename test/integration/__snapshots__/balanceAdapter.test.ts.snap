// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute balance endpoint should handle empty address list 1`] = `
{
  "error": {
    "message": "Input, at 'addresses', must be a non-empty array of objects.",
    "name": "AdapterError",
  },
  "status": "errored",
  "statusCode": 400,
}
`;

exports[`execute balance endpoint should handle invalid address error responses 1`] = `
{
  "error": {
    "message": "Invalid Ethereum address: 0xInvalidAddress",
    "name": "AdapterError",
  },
  "status": "errored",
  "statusCode": 400,
}
`;

exports[`execute balance endpoint should handle invalid blockheight error responses 1`] = `
{
  "error": {
    "message": "Invalid blockHeight for address 0x103B66487784F6e3b4C5B2AcA92758198554C3e1: -1. It must be a non-negative integer.",
    "name": "AdapterError",
  },
  "status": "errored",
  "statusCode": 400,
}
`;

exports[`execute balance endpoint should return success for a single address 1`] = `
{
  "data": {
    "result": [
      {
        "address": "0x103b66487784f6e3b4c5b2aca92758198554c3e1",
        "balance": "14.61",
      },
    ],
  },
  "result": "14.61",
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 978347471111,
    "providerDataRequestedUnixMs": 978347471111,
  },
}
`;

exports[`execute balance endpoint should return success for multiple addresses 1`] = `
{
  "data": {
    "result": [
      {
        "address": "0x103b66487784f6e3b4c5b2aca92758198554c3e1",
        "balance": "14.61",
      },
      {
        "address": "0x742d35cc6634c0532925a3b844bc454e4438f44e",
        "balance": "0.78",
      },
    ],
  },
  "result": "15.40",
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 978347471111,
    "providerDataRequestedUnixMs": 978347471111,
  },
}
`;
